### DISCUSSION TITLE
Passport for service-to-service authentication & authorization

### DISCUSSION THREAD

**Discussion by vuryss** [symfony/symfony#42476] [Q&A] [answered]
I see that in 5.4 the security passport interface is getting deprecated [(here)](https://github.com/symfony/symfony/commit/810599d2f0fd19558d140c19bc50eb4dc0021a2e) and will be removed in 6.0. The original passport requires a user badge on initiation. In my case - I will have some OAuth2 client_credentials authorization for service to server authentication. This means that the operation is not done by a user, but an application. I will have both operations executed by users (User entities) and such executed by applications (OAuth2 client entities without attached user).
I made a custom passport - OAuthPassport which I initiated without the user requirement. Now that's deprecated, does anyone know a better way to configure the authentication & authorization having user as an optional piece in the puzzle?

**Comment by faizanakram99** [comment, marked as answer]
I wanted to write this as a comment, but unlike Stack overflow it is not possible in Github.

Would `Symfony\Component\Security\Http\Authenticator\Passport\SelfValidatingPassport` work instead of a custom passport ?

As an example

```php
new SelfValidatingPassport(
    new UserBadge(
        (string) $token, // token from oauth2 client
        //, optionally pass userLoader closure as second argument to verify token for example
    )
);
```

This passport has been specially created for skipping credentials check.

https://symfony.com/index.php/doc/current/security/authenticator_manager.html#self-validating-passport

**Reply by vuryss** [reply]
It might work, but validating the token as a user loader seems more like a workaround.
Also the user badge will have a getUser() method which requires a user interface. In my case I do not have a user, but an application as a client. I will have to generate fake user interfaces, which again sounds like a workaround :(

My solution now it way more elegant and clean with a OAuthPassport, OAuthPassportChecker listening on CheckPassportEvent and validating only OAuth passports. Also an OAuth token without a user (only place where exception is thrown on getUser() but other tokens already do that too, like the NullToken).

**Reply by lyrixx** [reply]
Hello,

>It might work, but validating the token as a user loader seems more like a workaround.

Actually, it's very common to do that, and it's not a workarround. It's done for this exact purpose.
For example, when authentication is like "api.example.com?token=123", the token hold the user account **and** and the credential. So if you found the token in the DB (or remote API, or whatever you want) it's OK. That's why it's an `SelfValidatingPasspost`.

> Also the user badge will have a getUser() method which requires a user interface. In my case I do not have a user, but an application as a client. I will have to generate fake user interfaces, which again sounds like a workaround :(

Actually, your Application must be an instance of `UserInterace`. That's how your application will be known is the security system. For example in a Voter, you want to get your application to know if it can access or not a resource.
```php
class Footer extends Voter
{
     protected function voteOnAttribute($permission, $project, TokenInterface $token)
     {
          // here you wan the application and associated scopes (if it exists)
          $app = $token->getUser();
     }
}

```

> Also an OAuth token without a user (only place where exception is thrown on getUser() but other tokens already do that too, like the NullToken).

[`NullToken::getUser()`](https://github.com/symfony/symfony/blob/1d196159fde010590b857059b5219922031c284b/src/Symfony/Component/Security/Core/Authentication/Token/NullToken.php#L34-L37) does not throw an exception. Did I miss something?

**Reply by vuryss** [reply]
> Actually, it's very common to do that, and it's not a workarround. It's done for this exact purpose.

User loader should return a UserInterface - I don't want a user, I want a server authenticated client, which is definitely different from a normal user of the system. In my case I have both normal users and server (application) clients and the difference between them is important. Saying that everything is a user is very limiting, at least in my scenario.

> Actually, your Application must be an instance of UserInterace. That's how your application will be known is the security system. For example in a Voter, you want to get your application to know if it can access or not a resource.

Tokens have roles, which are sufficient for checking whether you have access to a resource or not when you do not have a user.
OAuth2 clients have scopes, which I map to roles in Symfony. Based on that I can identify which system can access what resource.
Also as I have created a OAuthToken - I can fetch the client from the token if I want to check ownership of an resource so a specific client.
So here user can be completely optional, as I like it for the server to server communication.

> NullToken::getUser() does not throw an exception. Did I miss something?

My bad, it's the setUser method, but it's essentially a token without user, so it's working good so far.

**Reply by lyrixx** [reply]
> User loader should return a UserInterface - I don't want a user, I want a server authenticated client, which is definitely different from a normal user of the system. In my case I have both normal users and server (application) clients and the difference between them is important. Saying that everything is a user is very limiting, at least in my scenario.

in https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/Security/Core/User/UserProviderInterface.php#L68 it says it must returns a `UserInterface`, but it doesn't mean it has to be a human. It can by something else. For example, in my application it could be (depending of the context) an instance of User, Agent, ApiToken (bound to an organization), Stripe, CleverCloud. The last two are other "applications".

In Symfony, a "User" means "The thing that is authenticating / authenticated".

So basically, anything can be a user.

> Tokens have roles, which are sufficient for checking whether you have access to a resource or not when you do not have a user.

OK, it's fine (but a bit limited) to only rely only on the roles in the Token, but you can do that if you want.

Let's take GitHub as example:
* They have oAuth access token(I can login on example.com with my GitHub account). So in this case, when example.com call GitHub, they use my access token => I'm identified as a user in GitHub
* They have application oAuth (I can grant circleci to access my repository). So in this case, when circle will try to fetch my repository, they will use an application id + token => They are identified as an application in GitHub. And in this case, just the oauth scope are not enough => You don't want circle to have access to all private repositories on github. Only  the ones bound to the final User.

---

**Summary:**

* a `UserInterface` could be **anything**
* I strongly recommend you to **implements `UserInterface` in your `Application` class**
* if not the following code if totally fine:
    ```php
    return new SelfValidatingPassport(
        new UserBadge(
            'application',
            fn () => new InMemoryUser('application', null, ['ROLE_API_USER']),
        )
    );
    ```
    You can change / extract the "right" role from your DB in the UserProvider instead of passing a callback like I did.

---

I hope this is clearer how the security works in symfony 🤗

**Reply by vuryss** [reply]
Well I guess I have to workaround a few things to get it like that. Having a UserInterface which can be a User or a OAuth client is a bit confusing, but I guess that's what I have to work with.

Thanks for the detailed replies, I understand this is how Symfony works, I just hoped there is a more concise naming or strategy to authenticate applications which are not necessary users.

**Reply by vuryss** [reply]
I will mark this answer as accepted, but if anyone has a better way of handling this situations - I'm still watching :)

**Reply by lyrixx** [reply]
> Well I guess I have to workaround a few things to get it like that. Having a UserInterface which can be a User or a OAuth client is a bit confusing, but I guess that's what I have to work with.

IMHO you should have 2 user provider. one for the regular user, and another one for the application.
And I guess both authentication system are different (like on [github](https://docs.github.com/en/rest/overview/resources-in-the-rest-api#authentication))

**Reply by vuryss** [reply]
Well at this stage of my API I'm consolidating everything into OAuth2 to avoid confusion with different authentication methods. So I'm authenticating normal users with a password grant - which has a client (the mobile or web application) but explicitly references a user in the JWT token. They are the owners of the resources in the system, so no need of the extended OAuth2 flows. Also for the REST-based service to service communication (both between internal services of the application and external registered applications) I'm using the client_credentials grant. 

#### So in summary:
I have 2 entities:
- UserEntity (real person)
- ClientEntity (application - both internal or external)

And the following flows:
- Person -> Web Application -> API :: OAuth2 (password grant) -> UserToken(UserEntity, ClientEntity, roles)
- Internal service -> API :: OAuth2 (client_credentials grant) -> ClientToken(ClientEntity, roles)
- External APP -> API :: OAuth2 (client_credentials grant) -> ClientToken(ClientEntity, roles)

---
So basically yes - I have one Repository for Users and different for Clients, which could count as providers.
Also the API clients (web and mobile) are also clients in the system, but they are not allowed to authenticate as clients (with clients_credentials) but only as resource owners - with password grant.
